GCP: Cleanup and refactor provisioning scripts

Edwin-Pau opened the pull request:
All provisioning scripts for all VM's have been reviewed and cleaned up. This includes cleaning up log messages to make them consistent across all VM's, refactoring certain provisioning scripts for better readability, and fixing logging for the CAC VM's Startup scripts have been renamed to provisioning scripts to reflect the functionality of these scripts.

Commit:
1/4 GCP: cleanup log messages for CentOS WS

Non-functional changes. Cleaned up the style for logging messages so
they are more consistent with the rest of the Terraform logging
messages. Log filenames are now the same as Windows workstations.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
3/4 GCP: cleanup and refactor CAC startup scripts

Non-functional changes. Cleaned up the style for logging messages so
they are consistent with the rest of the Terraform logging messages.

Refactored each step of this script into separate functions for
better readability.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
2/4 GCP: cleanup log messages for Windows WS

Non-functional changes. Cleaned up the style for logging messages so
they are consistent with the rest of the Terraform logging messages.
Corrected minor typos.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
4/4 GCP: cleanup log messages for DC

Non-functional changes. Cleaned up the style for logging messages so
they are consistent with the rest of the Terraform logging messages.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: change log filenames

Log filenames have been changed to use "provisioning.log" so that it is
consistent across all VM's.

Fixed CAC logging as it was missing a -a flag, which meant the log was
being overwritten by the CAC installer.

There are now two CAC logs: one for VM provisioning like other VM's,
and another cac-install.log for the CAC installer.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
1/5 GCP: rename CAC startup script names

CAC startup scripts have been renamed to provisioning scripts and
all associated variables and names have been changed to better reflect
the functions of these scripts.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
2/5 GCP: rename Centos startup script names

Centos-gfx and Centos-std startup scripts have been renamed to
provisioning scripts and all associated variables and names have been
changed to better reflect the functions of these scripts.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
3/5 GCP: rename Windows startup script names

Win-gfx and Win-std startup scripts have been renamed to provisioning
scripts and all associated variables and names have been changed to
better reflect the functions of these scripts.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
4/5 GCP: rename DC startup script names

DC startup scripts have been renamed to provisioning scripts and all
associated variables and names have been changed to better reflect
the functions of these scripts.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
5/5 GCP: rename startup script names in .sample

The references to startup scripts in terraform.tfvars.sample files
have been changed to provisioning scripts so that the wording is
consistent with the rest of the scripts.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Edwin-Pau commented:
- Ready for code review, rebased and manually merged changes for CAC provisioning scripts.
- Deployment tested successfully on latest rebase.

shermanyin commented on modules/gcp/centos-gfx/centos-gfx-startup.sh.tmpl:
```diff
@@ -350,8 +346,8 @@ else
     install_idle_shutdown
-
-    log "--> Installation is completed !!!"
+    
```
is this a tab here? line should be empty

shermanyin commented on modules/gcp/centos-std/centos-std-startup.sh.tmpl:
```diff
@@ -45,75 +45,77 @@ retry() {
 
     # Exit if any of the required variables are missing
     if [[ -z "$PCOIP_REGISTRATION_CODE" || -z "$AD_SERVICE_ACCOUNT_PASSWORD" ]]; then
+        log "--> ERROR: Required variables are missing, exiting..."
```
Ideally we say which one is missing

shermanyin commented on modules/gcp/centos-gfx/centos-gfx-startup.sh.tmpl:
```diff
@@ -105,15 +106,15 @@ install_gpu_driver() {
     if [[ $(lspci -d '10de:*:0300' -s '.0' | wc -l) -gt 0 ]] && ! (modprobe --resolve-alias nvidia > /dev/null 2>&1)
     then
-        log "--> Start to install gpu driver ..."
+        log "--> Installing gpu driver..."
```
gpu -> GPU (similar comment for rest of PR)

shermanyin commented on modules/gcp/centos-gfx/centos-gfx-startup.sh.tmpl:
```diff
@@ -241,32 +240,32 @@ join_domain()
         fi
 
-        log "--> Joining the domain"
+        log "--> Joining the domain..."
```
could display domain_name here. same for the next file.

shermanyin commented on modules/gcp/centos-gfx/centos-gfx-startup.sh.tmpl:
```diff
@@ -241,32 +240,32 @@ join_domain()
 
-        log "--> Joining the domain"
+        log "--> Joining the domain..."
         if [[ -n "$OU" ]]
```
Seems like $OU is never defined. could double compare with the CAM script and maybe get rid of this in the future.

shermanyin commented on modules/gcp/centos-gfx/centos-gfx-startup.sh.tmpl:
```diff
@@ -276,20 +275,20 @@ join_domain()
         if [[ $exitCode -eq 0 ]]
         then
-            log "--> Joined Domain '${domain_name}' and OU '$OU'"
+            log "--> Joined Domain '${domain_name}' and OU '$OU'."
```
OU seems to be undefined - then nothing gets displayed here. don't need to display the OU unless you put in a check to see if OU exist first like the above if [[ -n "$OU" ]]

shermanyin reviewed (changes requested):
Reviewed this for now - don't you want to address AWS scripts for CentOS at the same time? they have a lot of similarities. 

Also the next commit seems to be out of order even though order doesn't matter.

shermanyin commented on modules/gcp/win-gfx/win-gfx-startup.ps1.tmpl:
```diff
@@ -267,22 +266,21 @@ function Install-Idle-Shutdown {
-    "################################################################"
-    "Join Domain"
-    "################################################################"
+    "--> Joining domain..."
```
could add domain name

shermanyin commented on modules/gcp/win-gfx/win-gfx-startup.ps1.tmpl:
```diff
@@ -308,16 +306,16 @@ function Join-Domain {
 
             if ($PSItem.FullyQualifiedErrorId -match "AddComputerToSameDomain,Microsoft.PowerShell.Commands.AddComputerCommand") {
-                "WARNING: Computer already joined to domain."
+                "--> WARNING: Computer is already joined to the domain."
```
add domain_name like below

shermanyin commented on modules/gcp/win-gfx/win-gfx-startup.ps1.tmpl:
```diff
@@ -387,13 +385,12 @@ Install-Idle-Shutdown
-"################################################################"
-"Restart Computer"
-"################################################################"
+"--> Restarting computer..."
```
can get rid of this

shermanyin reviewed (changes requested):
same comments applies to both win-gfx-startup.ps1.tmpl and the std version.

shermanyin commented on modules/gcp/cac-igm/cac-startup.sh.tmpl:
```diff
@@ -38,156 +40,189 @@ get_credentials() {
-    if [ $RC -eq 0 ] || [ $RETRIES -eq 0 ]; then
-        break
-    fi
+install_ldap_utils() {
```
i feel like this should below to wait for dc rather than 2 functions.

shermanyin commented on modules/gcp/cac-igm/cac-startup.sh.tmpl:
```diff
@@ -38,156 +40,189 @@ get_credentials() {
+
+    log "--> Running command: $INSTALL_DIR/cloud-access-connector install"
+    log "--> CAC install options:"
+    log "-t $${CAC_TOKEN} --accept-policies"
```
since we treat cac token and pcoip reg code as secrets, we shouldn't print them out.

shermanyin commented on modules/gcp/cac-igm/cac-startup.sh.tmpl:
```diff
@@ -38,156 +40,189 @@ get_credentials() {
+    # for logging and debug purposes.
+    log "--> Ensure domain ${domain_name} can be resolved..."
+    TIMEOUT=1200
+    until host ${domain_name}; do
```
just realized these 3 loops just continue if they timeout. we should write a log saying something like " time out but continue anyway" since there is no point just failing and stopping here.

shermanyin reviewed (changes requested):
i only looked at cac-igm, i assume the cac file has the same changes.  I also assume the blocks of code are just moved around and didn't try to go thru line by line to see what changed functionally.

shermanyin commented on modules/gcp/dc/new_domain_admin_user.ps1.tpl:
```diff
@@ -43,45 +43,45 @@ function Decrypt-Credentials {
-Write-Output "Creating new AD Domain Admin account ${account_name}..."
-Write-Output "================================================================"
-
+"================================================================"
```
Didn't you get rid of these lines in the other files? having said that, I think I like to have section headers look a little different rather than just "-->" so it's easier to spot where we're at in the logs

Edwin-Pau replied:
I think I ended up keeping these lines and just removed Write-Output for consistency sake.

shermanyin commented on modules/gcp/dc/new_domain_admin_user.ps1.tpl:
```diff
@@ -43,45 +43,45 @@ function Decrypt-Credentials {
         if ($Elapsed -ge $Timeout) {
-            Write-Output "Error: Timed out trying to create new AD acccount."
-            exit
+            Write-Output "ERROR: Timed out trying to create new AD acccount, exiting..."
```
drop Write-Output?

Edwin-Pau replied:
I don't think we need Write-Output, will be dropping it. It'll just log it and then exit. I used exit 1 as the exit code as that's what everything else uses.

`Sends the specified objects to the next command in the pipeline. If the command is the last command in the pipeline, the objects are displayed in the console.`

shermanyin commented on modules/gcp/dc/new_domain_admin_user.ps1.tpl:
```diff
@@ -43,45 +43,45 @@ function Decrypt-Credentials {
-            Write-Output "Error: Timed out trying to create new AD acccount."
-            exit
+            Write-Output "ERROR: Timed out trying to create new AD acccount, exiting..."
+            exit 1
```
how does exit behave in powershell? I think it took the last command's return code?  does using Write-Output mask the exit code? do we even need the exit line?

shermanyin commented on modules/gcp/dc/new_domain_users.ps1.tpl:
```diff
@@ -26,13 +24,12 @@ foreach ($User in $ADUsers)
-    }
-    else
-    {
+        Write-Warning "WARNING: A user account with username $Username already exist in Active Directory."
```
if we use Write-Warning for warning, is normal output for Error ok or is there a Write-Error?  Will these affect how they show / not show in our logs?

shermanyin commented on modules/gcp/cac-igm/cac-startup.sh.tmpl:
```diff
@@ -5,8 +5,10 @@
+
 INSTALL_DIR="/root"
-INSTALL_LOG="$INSTALL_DIR/cac-install.log"
+CAC_INSTALL_LOG="$INSTALL_DIR/cac-install.log"
```
I think for convenience, might as well put the logs in provisioning.log too.

shermanyin commented on modules/gcp/cac-igm/cac-startup.sh.tmpl:
```diff
@@ -205,10 +207,17 @@ install_cac() {
+if [[ ! -f "$LOG_FILE" ]]
+then
+    mkdir -p "$(dirname $${LOG_FILE})"
+    touch "$LOG_FILE"
```
don't think you have to touch it by default. echo should create the file and it should already be 644 mode. I think all you have to do here is to make sure the directory exist.

You could also log "Script started" just to give the timestamp of when this is run.

Commit:
GCP: Cleanup CentOS provisioning scripts

Added domain_name and removed organizational unit from the log messages.
Fixed GPU capitalization and removed a tab from a line.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Fix logging for CentOS scripts

Moved the CAC install log to the same directory as the provisioning log.
Added a time stamp at the start of the logging file for the CAC
provisioning log. Removed unnecessary $$ escaping for Terraform
templating.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Cleanup CAC provisioning scripts

Moved ldap utils install function to the end of wait_for_dc function.
Added more logging messages when timeout has been reached inside
the wait_for_dc function.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Cleanup DC provisioning scripts

Removed the use of Write-Output and Write-Warning for logging as it is
not needed. Cleaned up the logging for new_domian_users script. Used
"exit 1" as the same terminating exit code as other Powershell scripts.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Cleanup Windows-WS provisioning scripts

Added section headers for better readability in the logs. Removed
unnecessary logging. Added missing domain_name in the logs.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Remove secrets from logs and check vars

Removed logging of any secrets during provisioning. Added explicit
checks for the required secrets to pinpoint if and which variables
are missing during the provisioning process.

Signed-off-by: Edwin-Pau <epau@teradici.com>

shermanyin commented on modules/gcp/centos-gfx/centos-gfx-provisioning.sh.tmpl:
```diff
@@ -69,8 +69,18 @@ get_credentials() {
-    if [[ -z "$PCOIP_REGISTRATION_CODE" || -z "$AD_SERVICE_ACCOUNT_PASSWORD" ]]; then
-        log "--> ERROR: Required variables are missing, exiting..."
+    if [[ -z "$PCOIP_REGISTRATION_CODE" ]]; then
+        log "--> ERROR: Missing <PCOIP_REGISTRATION_CODE> variable..."
```
Typically error messages shouldn't talk about the variable.  Should read more like: "Missing PCoIP Registration Code."

Note not "..." as it's not starting to do something like "installing xyz..."

shermanyin commented on modules/gcp/centos-gfx/centos-gfx-provisioning.sh.tmpl:
```diff
@@ -69,8 +69,18 @@ get_credentials() {
+    fi
+
+    if [[ -z "$AD_SERVICE_ACCOUNT_PASSWORD" ]]; then
+        log "--> ERROR: Missing <AD_SERVICE_ACCOUNT_PASSWORD> variable..."
```
"Missing Active Directory Service Account Password."

shermanyin commented on modules/gcp/centos-gfx/centos-gfx-provisioning.sh.tmpl:
```diff
@@ -69,8 +69,18 @@ get_credentials() {
-        log "--> ERROR: Required variables are missing, exiting..."
+    if [[ -z "$PCOIP_REGISTRATION_CODE" ]]; then
+        log "--> ERROR: Missing <PCOIP_REGISTRATION_CODE> variable..."
+        missing_vars=true
```
be consistent and use double quotes for strings?  missing_vars="true"

shermanyin commented on modules/gcp/cac-igm/cac-provisioning.sh.tmpl:
```diff
@@ -60,8 +60,23 @@ get_credentials() {
-    if [[ -z "$PCOIP_REGISTRATION_CODE" || -z "$AD_SERVICE_ACCOUNT_PASSWORD" || -z "$CAC_TOKEN" ]]; then
-        log "--> ERROR: Required variables are missing, exiting..."
+    if [[ -z "$PCOIP_REGISTRATION_CODE" ]]; then
+        log "--> ERROR: Missing <PCOIP_REGISTRATION_CODE> variable..."
```
same as the centos comments here

shermanyin commented on modules/gcp/cac-igm/cac-provisioning.sh.tmpl:
```diff
@@ -127,6 +121,7 @@ wait_for_dc() {
         -w $AD_SERVICE_ACCOUNT_PASSWORD \
         -o nettimeout=1; do
         if [ $TIMEOUT -le 0 ]; then
+            log "--> Timed out waiting for AD account ${ad_service_account_username}@${domain_name} to become available. Continuing..."
```
this should be at least an  ERROR

shermanyin commented on modules/gcp/cac-igm/cac-provisioning.sh.tmpl:
```diff
@@ -144,9 +139,10 @@ wait_for_dc() {
     TIMEOUT=1200
     until host ${domain_name}; do
         if [ $TIMEOUT -le 0 ]; then
+            log "--> Timed out trying to resolve ${domain_name}. Continuing..."
```
ERROR again

shermanyin commented on modules/gcp/cac-igm/cac-provisioning.sh.tmpl:
```diff
@@ -156,13 +152,33 @@ wait_for_dc() {
     TIMEOUT=1200
     until netcat -vz ${domain_name} 636; do
         if [ $TIMEOUT -le 0 ]; then
+            log "--> Timed out trying to contact ${domain_name}:636. Continuing..."
```
ERROR

shermanyin commented on modules/gcp/cac-igm/cac-provisioning.sh.tmpl:
```diff
@@ -156,13 +152,33 @@ wait_for_dc() {
         TIMEOUT=$((TIMEOUT-10))
         sleep 10
     done
+
+    # Wait for service account to be added. Do this last because it takes
+    # a while for new AD user to be added in a new Domain Controller.
+    # Note: using the domain controller IP instead of the domain name for
+    #       the host is more resilient.
+    log "--> Installing ldap_utils..."
+    RETRIES=5
+    while true; do
+        apt-get -qq update
+        apt-get -qq install ldap-utils
+        RC=$?
+        if [ $RC -eq 0 ] || [ $RETRIES -eq 0 ]; then
+            break
+        fi
+
+        log "--> ERROR: Failed to install ldap-utils. $RETRIES retries remaining..."
+        RETRIES=$((RETRIES-1))
+        sleep 5
+    done
```
this probably won't work? ldapwhoami above requires ldap-utils , this section has to come first in wait_for_dc()

shermanyin commented on modules/gcp/cac/cac-provisioning.sh.tmpl:
```diff
@@ -60,8 +60,23 @@ get_credentials() {
 check_required_vars() {
-    if [[ -z "$PCOIP_REGISTRATION_CODE" || -z "$AD_SERVICE_ACCOUNT_PASSWORD" || -z "$CAC_TOKEN" ]]; then
-        log "--> ERROR: Required variables are missing, exiting..."
+    if [[ -z "$PCOIP_REGISTRATION_CODE" ]]; then
```
same comments from the other cac provisioning script applies to this file.

shermanyin commented on modules/gcp/dc/new_domain_users.ps1.tpl:
```diff
@@ -7,6 +7,11 @@
+
+$LOG_FILE = "C:\Teradici\provisioning.log"
+
+Start-Transcript -path $LOG_FILE -append
```
can you please make sure Start-Transcript will document which file the code is from?  If it doesn't already write it, please output the name of this file so we know where the logs are from.

Edwin-Pau replied:
Confirmed that Start-Transcript will provide a header with a bunch of meta data including which file the code is from:
```
**********************
Windows PowerShell transcript start
Start time: 20200420205253
Username: VM-DC\Administrator
RunAs User: VM-DC\Administrator
Configuration Name: 
Machine: VM-DC (Microsoft Windows NT 10.0.17763.0)
Host Application: powershell -file C:/Temp/dc-provisioning.ps1
Process ID: 4200
PSVersion: 5.1.17763.1007
PSEdition: Desktop
PSCompatibleVersions: 1.0, 2.0, 3.0, 4.0, 5.0, 5.1.17763.1007
BuildVersion: 10.0.17763.1007
CLRVersion: 4.0.30319.42000
WSManStackVersion: 3.0
PSRemotingProtocolVersion: 2.3
SerializationVersion: 1.1.0.1
**********************
```

Commit:
GCP: Fix provisioning scripts

For the CentOS provisioning scripts, the logic to check for required
variables are now in a separate function called check_required_vars
just like the CAC provisioning scripts.

For the CAC and CentOS provisioning scripts, the check_required_vars
function has been updated to use double quotes for the "true" string,
representing the true boolean value for consistency.

Fixed logging messages in the above scripts when dealing with missing
required variables as well as error messages.

Moved the installation of ldap_utils to the beginning of the wait_for_dc
function.

Minor grammar fix in a DC script.

Signed-off-by: Edwin-Pau <epau@teradici.com>

shermanyin commented on modules/gcp/centos-gfx/centos-gfx-provisioning.sh.tmpl:
```diff
@@ -67,19 +67,20 @@ get_credentials() {
     fi
 
-    if [[ $missing_vars = true ]]; then
+    if [[ "$missing_vars" = "true" ]]; then
```
why do we need double quotes around the variable which is already a string?
